src/core/term.rs

```diff
@@ -139,4 +166,70 @@ impl<Witness, App: Application> UntypedProgram<Witness, App> {
     pub fn iter(&self) -> impl Iterator<Item = &Term<Witness, App>> {
         self.0.iter()
     }
+
+    /// Visit the DAG that underlies the program
+    /// and return a stack of node indices that can be popped such that the nodes are in post order.
+    fn get_post_order_stack(&self) -> Vec<usize> {
```

Comment:
In 1cef378b9dcf712fa46853b9d96d55e4c865a70a:

This would be a little more idiomatic and efficient as an iterator rather than a method that returns a complete `Vec`. But I like this algorithm.

Reply:
It would also be simpler to understand -- if we just yielded items on every loop iteration then we wouldn't have a list returned in the reverse order and a `post_order` method being called from a method that needs a pre ordering.

Reply:
The same comment applies to 41ae92f368f222738c1b99058f11f723b86e1155 where you implement essentially the same algorithm on `TermDag`.

I also wonder whether we could abstract these into a trait where you have to implement `get_left` and `get_right` and then get the iteration logic for free. In `TermDag` we iterate over `&Self`s and here we iterate over `usize`s but it's easy to make this type an associated type of the trait.

Reply:
That's a great idea; I will write an iterator.

Reply:
We now have the trait `DagLike` that comes with a default implementation of a post-order iterator.